php/libraries/NDB_BVL_Instrument.class.inc

```diff
@@ -2047,6 +2047,31 @@ class NDB_BVL_Instrument extends NDB_Page
         return $defaults[0];
     }
 
+    /**
+     * Gets specified fields of data from an instrument out of the database and
+     * returns it as an array.
+     *
+     * @param NDB_BVL_Instrument $instrumentInstance The object whose data is
+     *                                               to be retrieved.
+     * @param array $fieldsArray Contains the fields of interest
+     *
+     * @return array containing the data for the fields specified in $fieldsArray
+     */
+    static function loadInstanceDataFromArray($instrumentInstance, $fieldsArray)
```

Comment:
I don't see what you're gaining by this. You're loading all the data anyways, and then just discarding it.

Reply:
it would be to reduce duplication of code when a only subset of fields are needed especially in scoring functions.

Reply:
Also since the cost of a select is the same whether we fetch all fields or not, calling loadInstanceData also reduces the redundancy of having to check whether the instrument is using jsonData or not.

Reply:
The cost of a SELECT isn't the same whether you fetch all fields or not. See https://blog.jooq.org/2017/03/08/many-sql-performance-problems-stem-from-unnecessary-mandatory-work/ .. I still don't see how this reduces code duplication. In what circumstances would this reduce duplication as opposed to just ignoring the fields you don't care about?

(re: the question about major/minor, the release branch is pushed into both of them now, so it's no longer relevant..)

Reply:
@driusan in almost all our instruments that have scoring, we fetch the data saved in the table and use the data to run the score, except the score fields and the data are in the same table !! what we used to do is to select the specific fields necessary for scoring (ie `SELECT A,B,C,D ... FROM instruement`) and then run a for loop that does A + B + C + D.


In order tostart using the JsonData we cant run direct query databases anymore, instead we would need to load the instance data and either run through and unset the score fields and then calculate score OR in the same loop doing the scoring we add a `if(this field is not a scoring field) { do something}`

in either of the cases above that would be code that we would have to add to each and every scoring instrument. that function allows us to maintain status quo in a way where the code currently expecting a fixed array of fields would not have to be altered... the query would just be done using the new php function instead of a database query

whatcha think ??

Reply:
I still don't understand.. whether you're using `loadInstanceData` or you're using `SELECT A, B, C, D` you're still doing `$score = $row['A'] + $row['B'] + $row['C'] + $row['D'];`. `$row` containing other items that aren't used isn't going to affect that.

Reply:
when 50* fields contribute to the score and its the same in 20 other instruments, we dont do `$score = $row['A'] + $row['B'] + ...`, we do `SELECT A, B, C, D` and then `foreach ($rows as row) { sum this and that}`.

1. because that the way it's currently implemented and so changing that at this point is way more costly then just putting a wrapper around `loadInstanceData`
2. because we reuse the same array multiple times in different ways for different scoring algorithms... so might as well get the entire array just once and loop through

Reply:
@driusan final thoughts ??

Either way we would copy the changes into the CCNA instrument extension classes if LORIS doesnt need it... so no waste or anything

Reply:
I've never heard of anyone else doing SELECT A, B, C and then foreaching on the columns of the row (in most scoring algorithms I've seen it's more complex than that anyways since there are usually a couple columns that are reverse-coded or have a lookup or something.)

Do you know of any other projects doing that? If so, I'm not strictly opposed to it, it just seems unnecessary to me based on my experience.

Reply:
I agree with Rida, will most likely copy this over to CCNA instrument extensions if LORIS doesn't want it. Otherwise a huge refactor of scoring will be needed and I'm not willing to go  down that road...